File: packages/app/src/services/software/definitions/wireshark.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'wireshark';
	public name = 'Wireshark';
	public category = [SoftwareCategory.Network, SoftwareCategory.Utility];
	public downloadName = 'Wireshark-x64.exe';
	public shouldCacheUrl = true;
	public requiresAdmin = true;
	public icon = 'wireshark.png';
	public homepage = 'https://wireshark.org';

	public async resolveDownloadUrl() {
		const res = await fetch('https://wireshark.org');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/https:\/\/\d+.na.dl.wireshark.org\/win64\/Wireshark-\d+.\d+.\d+-x64.exe/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[0]);
	}
}

File: src/lib/software/wireshark.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class Wireshark(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'wireshark'
        self.name = 'Wireshark'
        self.category = [SoftwareCategory.Network, SoftwareCategory.Utility]
        self.download_name = 'Wireshark-x64.exe'
        self.is_archive = False
        self.should_cache_url = True
        self.requires_admin = True
        self.icon = 'wireshark.png'
        self.homepage = 'https://wireshark.org'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        pattern = compile(r'https://\d+.na.dl.wireshark.org/win64/Wireshark-\d+.\d+.\d+-x64.exe')
        html = reply.readAll().data().decode()
        match = pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(match.group(0))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest(self.homepage))
